File: leetcode/CopyListWithRandomPointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None
        
        # --- Pass 1: Interweave the original and new nodes ---
        # Iterate through the original list. For each node, create a copy
        # and place it immediately after the original node.
        # Original: A -> B -> C
        # After Pass 1: A -> A' -> B -> B' -> C -> C'
        curr = head
        while curr:
            # Create a new node with the same value as the current node.
            new_node = Node(curr.val)
            # Link the new node into the list.
            new_node.next = curr.next
            curr.next = new_node
            # Move to the next original node (skipping over the new one we just added).
            curr = curr.next.next
        
        # --- Pass 2: Set the random pointers for the new nodes ---
        # Now that every new node is next to its original, we can easily set the random pointers.
        curr = head
        while curr:
            # If the original node has a random pointer...
            if curr.random:
                # ...the new node's random pointer should point to the copy of the original's random target.
                # `curr.random.next` is the copied node corresponding to `curr.random`.
                curr.next.random = curr.random.next
            # Move to the next original node.
            curr = curr.next.next
        
        # --- Pass 3: Unlink the two lists ---
        # Separate the interwoven list into the original list and the new copied list.
        new_head = head.next
        curr_old = head
        curr_new = head.next

        while curr_old:
            # Restore the 'next' pointer of the original list.
            curr_old.next = curr_old.next.next
            # Set the 'next' pointer for the new list, if it's not the last node.
            if curr_new.next:
                curr_new.next = curr_new.next.next
            # Move to the next pair of old and new nodes.
            curr_old = curr_old.next
            curr_new = curr_new.next

        # Return the head of the newly created deep copy.
        return new_head

        

File: leetcode/medium/CopyListWithRandomPointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    /**
     * Copies a linked list with next and random pointers.
     * This method uses an O(1) space, three-pass approach to avoid a HashMap.
     * * Time Complexity: O(N), where N is the number of nodes. (Three passes over the list).
     * Space Complexity: O(1), excluding the space for the new list.
     *
     * @param head The head of the original linked list.
     * @return The head of the new, deep-copied linked list.
     */
    public Node copyRandomList(Node head) {
        if (head == null) {
            return null;
        }

        // --- Pass 1: Interweave the new nodes into the original list ---
        // For each original node 'curr', create a copy and place it right after 'curr'.
        // Original: N1 -> N2 -> N3
        // After Pass 1: N1 -> N1_copy -> N2 -> N2_copy -> N3 -> N3_copy
        Node curr = head;
        while (curr != null) {
            Node copy = new Node(curr.val);
            copy.next = curr.next; // copy.next points to N(i+1)
            curr.next = copy;      // curr.next now points to N(i)_copy
            curr = copy.next;      // Advance 'curr' to the original next node N(i+1)
        }

        // --- Pass 2: Set the 'random' pointers for the new nodes ---
        // If an original node 'curr' has a random pointer to 'R', then its copy 
        // 'curr.next' must have a random pointer to 'R.next' (which is R_copy).
        curr = head;
        while (curr != null) {
            Node copy = curr.next;
            if (curr.random != null) {
                // curr.random is the original node R. R.next is the copy node R_copy.
                copy.random = curr.random.next; 
            }
            // Advance to the next original node (skip over the copy)
            curr = curr.next.next; 
        }

        // --- Pass 3: Separate the original and new lists and restore the original list ---
        Node newHead = head.next;
        Node newCurr = newHead;
        curr = head;

        while (curr != null) {
            // Restore original link: N(i) -> N(i+1)
            curr.next = newCurr.next;
            curr = curr.next; // Advance original pointer

            if (curr != null) {
                // Separate new list link: N(i)_copy -> N(i+1)_copy
                newCurr.next = curr.next;
                newCurr = newCurr.next; // Advance new pointer
            }
        }

        return newHead;
    }
}